resolve_fleet_files leaks temp files on error exit

If `resolve_fleet_files` fails mid-loop (e.g. a bad ref at index 3 of 5), temp files written for earlier inline agents are left in FLEET_TMPDIR. The caller's `|| exit 1` exits before `cleanup_fleet_tmpdir` is called. Add a `trap cleanup_fleet_tmpdir EXIT` in apply.sh, plan.sh, and status.sh to guarantee cleanup even on error.

_Follow-up from #31_

# Implementation Plan

I now have all the context needed to write a precise implementation plan.

---

# Implementation Plan: Issue #158 — `resolve_fleet_files` temp file leak on error exit

## 1. Objective

Add `trap cleanup_fleet_tmpdir EXIT` to `apply.sh`, `plan.sh`, and `status.sh` so that `FLEET_TMPDIR` is always removed — even when `resolve_fleet_files` returns early with an error (e.g. bad ref at index 3 of 5) and the caller's `|| exit 1` fires before the explicit `cleanup_fleet_tmpdir` call.

The change targets the **issue-31 worktree** (`/home/mvillmow/Projects/Myrmidons/.worktrees/issue-31/`) where fleet support lives, not the main branch (which has no fleet code yet).

---

## 2. Approach

### Key decisions

- **Register the trap before `resolve_fleet_files` is ever called** — the critical ordering rule from team knowledge. If the trap is registered after `FLEET_TMPDIR` is created (line 130 of `reconcile.sh`), a failure in the very first `mktemp -d` call window is still covered. The safe position is at the top of `main()`, before `find_fleet_file` / `resolve_fleet_files`.
- **Keep the explicit `cleanup_fleet_tmpdir` calls in place** — the `EXIT` trap fires on every exit path including the happy path, so double-calling `cleanup_fleet_tmpdir` is harmless (it guards with `[[ -n "${FLEET_TMPDIR:-}" ]]`). Leaving explicit calls preserves readability and intent.
- **Do not silence cleanup errors** — per team knowledge (`credential-mount-context-manager`), bare `|| true` in cleanup masked 498 leaked dirs. `cleanup_fleet_tmpdir` already uses `rm -rf` which does not error on missing paths; no additional silencing needed.
- **Scope**: three files in `.worktrees/issue-31/scripts/` — `apply.sh`, `plan.sh`, `status.sh`. No changes to `reconcile.sh` (the cleanup function is already correct).

---

## 3. Files to Create

| File | Purpose |
|------|---------|
| `.worktrees/issue-31/tests/test-fleet-tmpdir-trap.sh` | BATS-style test (plain bash, matching existing test style) — simulates a bad ref at index 3 of 5, asserts temp dirs from indices 0–2 are removed after the script exits non-zero |

---

## 4. Files to Modify

### `.worktrees/issue-31/scripts/apply.sh`

**Where:** Inside `main()`, after `agamemnon_check_connection` and before the `if [[ -n "$FLEET" ]]` block (line ~88).

**Change:** Add one line:
```bash
trap cleanup_fleet_tmpdir EXIT
```

**Before** (line ~88):
```bash
    local yaml_files
    if [[ -n "$FLEET" ]]; then
```

**After:**
```bash
    trap cleanup_fleet_tmpdir EXIT
    local yaml_files
    if [[ -n "$FLEET" ]]; then
```

### `.worktrees/issue-31/scripts/plan.sh`

**Where:** Inside `main()`, after `agamemnon_check_connection` and before the `if [[ -n "$FLEET" ]]` block (line ~58).

**Change:** Add one line:
```bash
trap cleanup_fleet_tmpdir EXIT
```

**Before** (line ~58):
```bash
    local yaml_files
    if [[ -n "$FLEET" ]]; then
```

**After:**
```bash
    trap cleanup_fleet_tmpdir EXIT
    local yaml_files
    if [[ -n "$FLEET" ]]; then
```

### `.worktrees/issue-31/scripts/status.sh`

**Where:** Inside `main()`, after `agamemnon_check_connection` and before the `if [[ -n "$FLEET" ]]` block (line ~53).

**Change:** Add one line:
```bash
trap cleanup_fleet_tmpdir EXIT
```

**Before** (line ~53):
```bash
    local yaml_files
    if [[ -n "$FLEET" ]]; then
```

**After:**
```bash
    trap cleanup_fleet_tmpdir EXIT
    local yaml_files
    if [[ -n "$FLEET" ]]; then
```

---

## 5. Implementation Order

1. **Edit `apply.sh`** — add `trap cleanup_fleet_tmpdir EXIT` in `main()` before fleet resolution block
2. **Edit `plan.sh`** — same change
3. **Edit `status.sh`** — same change
4. **Create `tests/test-fleet-tmpdir-trap.sh`** — write test that:
   - Creates a fleet YAML with 5 agents (3 inline, then a bad `ref:` at index 3)
   - Stubs out `agamemnon_*` functions (not needed for unit test)
   - Calls `resolve_fleet_files` directly with the trap in scope
   - Simulates an error exit (bad ref triggers `return 1` → caller `exit 1`)
   - Asserts that the temp dir created for the first 3 inline agents no longer exists
5. **Run the new test** — verify it passes with the trap in place; optionally verify it fails when the trap is removed (regression guard)

---

## 6. Verification

### Manual
```bash
cd .worktrees/issue-31

# Create a fleet with a bad ref at position 3 and 2 valid inline agents before it
# Run plan.sh with that fleet — it should exit 1 and leave no files in /tmp/myrm*
ls /tmp/myrm* 2>/dev/null || echo "clean — no leaked temp dirs"
```

### Automated test (new file)
```bash
./tests/test-fleet-tmpdir-trap.sh
```

Expected output:
```
--- trap cleanup: bad ref at index 3 ---
  PASS: FLEET_TMPDIR populated after inline agents
  PASS: temp dir removed after error exit via trap
=== Results: 2 passed, 0 failed ===
```

### Existing test regression
```bash
./tests/test-fleet-resolution.sh   # must still pass
```

### ShellCheck (optional, if CI runs it)
```bash
shellcheck -S warning scripts/apply.sh scripts/plan.sh scripts/status.sh
```
Note: `trap cleanup_fleet_tmpdir EXIT` with a function name (not a string) avoids SC2064 (expand-at-trap-time vs. define-time). No new warnings expected.

---

## 7. Skills Used

**Invoked during planning:**
- `hephaestus:advise` — searched team knowledge base for prior learnings on temp-dir leak patterns and trap registration ordering

**Team knowledge base skills referenced in Prior Learnings:**
- `credential-mount-context-manager` — confirmed trap/finally pattern is the correct fix; warned against silencing cleanup errors (`|| true`) and against placing resource creation before the trap is registered
- `bats-shell-testing` — BATS framework patterns for simulating failure exits and asserting cleanup behavior; note: existing tests in this repo use plain bash (not BATS runner), so the new test follows the same plain-bash style as `test-fleet-resolution.sh`
- `shellcheck-scope-templates` — SC2064 awareness for trap with function names vs. string expansion; no changes needed since `cleanup_fleet_tmpdir` is a defined function, not a variable expansion

---
*Generated by Claude Code Planner*

--- a/scripts/lib/reconcile.sh
+++ b/scripts/lib/reconcile.sh
@@ -187,6 +187,15 @@ resolve_fleet_files() {
     local repo_root
     repo_root="$(cd "$(dirname "${BASH_SOURCE[0]}")/../.." && pwd)"
 
+    # Trap to clean up any FLEET_TMPDIR we create on error exit (#158)
+    _resolve_fleet_cleanup() {
+        if [[ -n "${FLEET_TMPDIR:-}" && -d "${FLEET_TMPDIR}" ]]; then
+            rm -rf "${FLEET_TMPDIR}"
+            FLEET_TMPDIR=""
+        fi
+    }
+    trap '_resolve_fleet_cleanup' ERR
+
     local fleet_host
     fleet_host="$(yq eval '.metadata.host // ""' "$fleet_file")"
 
@@ -419,6 +428,8 @@ report_unmanaged() {
     # Guard: nothing to check if Agamemnon returned no agents
     [[ -z "$agents_json" || "$agents_json" == '[]' ]] && return 0
     shift
+    # Guard: nothing to check if no YAML files are managed (#130)
+    [[ $# -eq 0 ]] && return 0
     while IFS= read -r actual_name; do
         log_warn "[-] UNMANAGED ${actual_name} (in Agamemnon but not in desired state — use --prune to remove)"
     done < <(get_unmanaged_names "$agents_json" "$@")
@@ -432,15 +443,15 @@ get_unmanaged_names() {
     # Guard: nothing to check if Agamemnon returned no agents
     [[ -z "$agents_json" || "$agents_json" == '[]' ]] && return 0
     shift
+    # Guard: nothing to check if no YAML files are managed (#130)
+    [[ $# -eq 0 ]] && return 0
     local yaml_files=("$@")
 
-    # Collect all managed names from YAML files
+    # Collect all managed names from YAML files in a single yq invocation (#233)
     local managed_names=()
-    for yaml_file in "${yaml_files[@]}"; do
-        local n
-        n="$(yq eval '.metadata.name' "$yaml_file")"
-        managed_names+=("$n")
-    done
+    while IFS= read -r n; do
+        [[ -n "$n" ]] && managed_names+=("$n")
+    done < <(yq eval '.metadata.name' "${yaml_files[@]}")
 
     # Emit names present in Agamemnon but absent from managed list
     while IFS= read -r actual_name; do

